### ISSUE TITLE
Implementation of Custom Loss Function 

### ISSUE THREAD

**Issue by JandhyalaSri** [facebookresearch/fastText#1066]
Hello Team,
	We are trying to create our own loss function by extending the negative sampling loss function referring to loss.cc & loss.h . For extend of implementation we would like to know about the parameters utilized. Also, I am new to C++ . So,Can we get any references/description for loss class.

@Celebio Do you have any suggestions how do i implement this. That would be a great help .

Thanks in advance.

**Comment by dongrixinyu** [comment]
I have the same problem. I wanna apply focal loss to fasttext in order to solve unbalanced dataset

**Comment by Celebio** [comment]
Hi @JandhyalaSri ,
Indeed you need to write a class that inherits from `Loss` class that is, as you mentioned, described in [the file `loss.h`](https://github.com/facebookresearch/fastText/blob/master/src/loss.h#L23). 

Your class will receive the output matrix as argument in the constructor. In the `forward` function, if `backprop` argument is true, your class will have the responsibility to update the output matrix accordingly to the backpropagation, as well as the hidden layer's gradient that is in `state.grad`.

You can have a look in the [implementation of SoftmaxLoss](https://github.com/facebookresearch/fastText/blob/master/src/loss.cc#L322) to get inspired.

I hope it answers your question. I will close the issue now. Feel free to re-open it if you have further questions.

Best regards,
Onur
